doc/sphinx/proofs/writing-proofs/equality.rst

```diff
@@ -640,6 +640,8 @@ which reduction engine to use.  See :ref:`type-cast`.)  For example:
 
 .. tacn:: red @simple_occurrences
 
+   .. todo PR: need better definition of head constant with an example or two
```

Comment:
If a term is of the form `c t__1 ... t__n`, where `c` is a constant, then `c` is the head constant. I'm not sure this is the best place for this definition, but I don't have any suggestion, so better define it here than nowhere.

Reply:
And in general, `t__1` is the head of `t__1 ... t__n`?  "constant" meaning not an expression or not a variable?

I put this:

```
   In the term :n:`t__1 ... t__n`, :n:`t__1` is the :gdef:`head` of the term.
   In a term with the form :n:`c t__1 ... t__n`, where :n:`c` is constant,
   :n:`c` is the :gdef:`head constant`.  
```

Reply:
Ah, this doesn't make complete sense.  It seems to suggest that `forall ..., T` is replaced by `T`, but the `forall` isn't removed, is it?  And AFAIU `c t__1 ... t__n` is just a different way of describing the term `T`, so almost a no-op.

![image](https://user-images.githubusercontent.com/1253341/184706280-ed6922a2-8e36-48f9-a310-e0c5997121a0.png)

Reply:
Constant is already a defined term: https://coq.inria.fr/refman/language/core/definitions.html#term-constant

> And in general, `t__1` is the head of `t__1 ... t__n`? "constant" meaning not an expression or not a variable?

Yes, but only if `t__1` itself is not an application.

Regarding your second comment, I am lacking context.

Reply:
In `(foo bar) 1 2 3`, the head is an application?  You would not generally call it a head because you have to deal with "foo bar" first??

For the second question, consider this (`red` on line 74 in `Sets/Image.v').  `c` in `T` is `Included`.  The `c t__1 ... t__n` is about the term _before_ the change rather than after.  Right?

Before:
![image](https://user-images.githubusercontent.com/1253341/184720274-e2a32f92-13ee-4685-887e-f12f117624b3.png)

After:
![image](https://user-images.githubusercontent.com/1253341/184720316-2eeb18fa-c999-4b13-91f2-63af948d2d41.png)

Reply:
> In `(foo bar) 1 2 3`, the head is an application? You would not generally call it a head because you have to deal with "foo bar" first??

`(foo bar) 1 2 3` is equivalently written `foo bar 1 2 3` (using the default associativity rule) thus the head is `foo`. (The head is never an application.)

Reply:
> The `c t__1 ... t__n` is about the term _before_ the change rather than after. Right?

Reading the current description, it looks like it designate the result of the reduction. But your example clearly shows that this description is incomplete, since there are cases where this reduction makes a forall appear.

Reply:
I'm a bit unclear on what an application is. Can you give an example of a term starting a an application?

Reply:
`a b` is what is called an application.

My point is just that if you have `(foo bar) 1 2 3`, you do not call `foo bar` the head.

Reply:
Is it useful to define "head"?  Check what I wrote.